<a href="https://pandas.pydata.org/">
<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/ed/Pandas_logo.svg/2560px-Pandas_logo.svg.png" width="300px">
</a>


# Data Wrangling with pandas



## Objectives
This is the demonstration notebook for the first task. The facilitators will present a lecture that is based around this notebook. Before performing the first task as a team, we recommend you first go through this walkthrough together, as it will show a lot of techniques and code that will be useful in the task.  


The main objective of this walkthrough is to familiarize you with the [pandas python package](https://pandas.pydata.org/). Pandas is a widely-used popular Python package that can discover, clean and structure tabular data. In particular, this notebook will cover:
* The basic concept behind the core data structures within Pandas: [Series](https://pandas.pydata.org/docs/user_guide/dsintro.html#series) and [DataFrames](https://pandas.pydata.org/docs/user_guide/dsintro.html#dataframe)
* How to process, organize and clean data stored in DataFrames
* Visualization of data stored within a DataFrame.

After completing this notebook you will be able to:
* Understand key features of the Pandas library.
* Import tabular data from spreadsheets into Pandas
* Understand your data, its size, features and its structure.
* Handle missing values
* Correct data format
* Transform variables, including grouping continuous values and normalization
* Perform statistical analysis
* Visualise data using Matplotlib library


## What do we mean by data wrangling?
Data wrangling is the process of reading in the raw data and then cleaning and processing it so that the resulting information is in suitable format for analysis. Part of the data wrangling process includes data discovery, data cleaning (identifying and then removing any erroneous data), handling missing data, transforming variables, and visualizing the resulting data.

Data wrangling is often needed before any quantitative analysis can begin; however, it is often the most time-consuming and tedious part of the process. There are many different skills and actions that are required as part of data wrangling, many of which are illustrated in the example code below. Many of these steps are part of the functionality that Pandas provides. 

Here are some common, often used steps that are considered part of data wrangling:

* **Discovering**: 
Reviewing and understanding the input data to better understand its structure and what variables will be useful for your problem 
* **Structuring**:
Standardising the format for disparate types of data and make the data usable for automated or semiautomated data analysis. The data must be structured to fit the analytics model 
COME BACK TO THIS

* **Cleaning** : 
There are often missing or implausible values in a dataset. These occur for a number of reasons and they could adversely affect the reliability and repeatability of an analysis. Cleaning techniques are identify and in some cases correct these instances so that they can be handled appropriately in subsequent analyses.

* **Enriching**
You are pretty familiar with the data at this point. This is the moment to ask yourself if you want to enhance the data. Are you looking to add other data to it?


*  **Validating** 
This step involves iterative programming steps that authenticate your data’s quality and safety. For example, you may have problems if your data is not clean or enriched and the attributes are not distributed evenly.

Many of these different stages of data wrangling require you to visualize the data in some form other than the base tabular data. Visualisation libraries such as matplotlib and seaborn are often used to plot data as well as perform some basic statistical analysis. For example, one simple way of identify outliers for cleaning is to plot the data points and visually identify points that appear to be well outside the plausible range of the rest of the data for a given variable. 



## Table of Contents
* [1. Importing pandas and reading in data](#1-importing-pandas-and-reading-in-data)
* [2. Discovering and reviewing the data](#2-discovering-and-reviewing-the-data)
* [3. Handling missing data](#3-handling-missing-data)
* [4. Checking the data representation](#4-checking-the-data-representation)
    <li><a href="https://#data_standardization">Data standardization</a></li>
    <li><a href="https://#data_normalization">Data normalization (centering/scaling)</a></li>


## 1. Importing pandas and reading in data
### 1.1 Importing packages
The first thing that we need to do is to import the Python packages we are going to need. This includes pandas, [numpy](https://numpy.org/) (which Pandas uses to store data in the two-dimensional tabular structure), and [matplotlib](https://matplotlib.org/), a popular Python package used for data visualisation.

For convenience, we will import these packages with the widely agreed shortname (`pd`,`np`,`plt`), so that we don't always have to provide the package name each time we use some functionality from the package. 

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt


###  1.2 Reading in data
For this notebook, we are using a dataset provided by the World Health Organisation. There are many sources where you might get data from that you will then want to work with using pandas. The most common is to open a file containing tabular data. The most well-known format for tabular data are Excel spreadsheets. Another commonly used format are plain-text *comma separated values* files with the extension `csv`. Each line of a csv file represents a row, and each column is separated by a comma. 

Besides local files, you may also get data by querying a database or requesting data from the web. 

All of these sources generate tabular data that can be stored within pandas DataFrames.

For the purpose of this notebook, we have assumed you have [downloaded the data](https://liveuclac.sharepoint.com/:x:/r/sites/TeamCodersEventsPlanning/Shared%20Documents/General/WHO/who_case_statistics_modified3.csv?d=wec3e29c1cb9d43779f37cc43bf9eed3b&csf=1&web=1&e=wy7e84) and stored it to a local file in the same directory as this notebook. 

pandas has many different functions that will read in data depending on the data format. Below, we will read the filename and store it the variable `input_file` and then use the [read_csv](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) function to read this file 
and store it into a DataFrame, under a variable named `df_raw`. Here the df stands for DataFrame, so that you know what is stored in the variable.  

In [3]:

# If you have not downloaded the file into this directory
# please ammend the first line so that input_file contains
# the full location (path) of the file
input_file = "data/who_case_statistics_modified3.csv" 
df_raw = pd.read_csv(input_file)

## 2. Discovering and reviewing the data 
Before we can work with the data, we need to know about more about the contents of the data. Here are some core functions that can help you start to review the structure and contents of your DataFrame. 
* `pandas.DataFrame.head()` - prints out the first *n* rows of the DataFrame 
* `pandas.DataFrame.tail()` - prints out the last *n* rows of the DataFrame

You can understand a bit more of the data structure using the following commands:
* `pandas.DataFrame.info()` - summary information (column names, counts, data types) 
* `pandas.DataFrame.describe()` - provides basic summary statistics for each variable

Some basic low-level information about the DataFrame can be extracted with the following attributes:
* `pandas.DataFrame.shape` - provides the number of rows and columns in the DataFrame
* `pandas.DataFrame.dtypes` - indicates the datatype (string, float, integer, etc) used to store the data for each variable


### 2.1 Review the data
Use the method <b>head()</b> to display the first five rows of the dataframe.


In [37]:
df_raw.head()

country  year     sex          age cases_no  population  HDI for year  \
0  Albania  1987    male  15-24 years       21      312900           NaN   
1  Albania  1987    male  35-54 years       16      308000           NaN   
2  Albania  1987  female  15-24 years       14      289700           NaN   
3  Albania  1987    male    75+ years        1       21800           NaN   
4  Albania  1987    male  25-34 years        9      274300           NaN   

   gdp_for_year ($)   
0      2,156,624,900  
1      2,156,624,900  
2      2,156,624,900  
3      2,156,624,900  
4      2,156,624,900

In [38]:
df_raw.shape

(27840, 8)

In [4]:
df_raw.dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27840 entries, 0 to 27839
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   country             27840 non-null  object 
 1   year                27840 non-null  int64  
 2   sex                 27840 non-null  object 
 3   age                 27840 non-null  object 
 4   cases_no            23575 non-null  object 
 5   population          27840 non-null  int64  
 6   HDI for year        8368 non-null   float64
 7    gdp_for_year ($)   27840 non-null  object 
dtypes: float64(1), int64(2), object(5)
memory usage: 1.7+ MB


### Exercise 1
Use the tail method to see the last part of the data.
How many lines do you see?
Use the the tail method again to see the last 20 lines of the data.

In [ ]:
# Write your code here

### Exercise 2
Use the info method to gather information  about the data. 


In [ ]:
# Write your code here

### Observations
* The data frame contains 27840 rows and 8 columns.
* The data frame is an object of class pandas.DataFrame.
* Columns that hold only numeric data are stored as floats or integers
* Columns that hold text or a mixture of data are stored as objects. 


### Exercise 3
Write any other observation that you see about the data. 


### 2.2 Understanding headers
How did pandas know how to name the columns? If you opened up the file in a text editor, you would see that the first row contains these names. When you run `read_csv`, pandas assumes that the column names (or _headers_) are located in the first row. If that is not the case, you can specify which row they can be found by using the `headers` argument. 

However, if there are not any column names in the file, then it is important to tell pandas that, lest you lose actual data. This can be done by setting `headers` to `None`. 


In [ ]:
# Reading in the data - but assuming no headers
df_raw = pd.read_csv(input_file, headers=None)
# Now look at the head, does it look correct?
df_raw.head()

In [ ]:
# Also look at the info, does it look correct?
df_raw.info()

### 2.3 Selecting the columns and getting the values
The way to to select columns is to write the 
`dataframe['column_name']`
Lets show all the countries in the data set.

In [5]:
df_raw = pd.read_csv(input_file)
countries=df_raw['country']
print(countries)

0            Albania
1            Albania
2            Albania
3            Albania
4            Albania
            ...     
27835        Belgium
27836       Thailand
27837    Netherlands
27838        Grenada
27839         Mexico
Name: country, Length: 27840, dtype: object


We can get the counts of different values in the countries column using the `pandas.DataFrame.value_counts()` function.

In [6]:
df_raw['country'].value_counts() #select the column with data_raw['country'] then add the method .value_counts()

Austria                   383
Netherlands               383
Iceland                   382
Mauritius                 382
Belgium                   373
                         ... 
Bosnia and Herzegovina     24
Macau                      12
Cabo Verde                 12
Dominica                   12
Mongolia                   10
Name: country, Length: 101, dtype: int64

This shows us how many times each country is found in the column. If you wanted instead to get a list of all countries found in this column, you can use the `unique` function. The output from this will be a one-dimensional numpy array.  


In [22]:
country_array = df_raw['country'].unique()
print(country_array)
print(type(country_array))

['Albania' 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Aruba' 'Australia'
 'Austria' 'Azerbaijan' 'Bahamas' 'Bahrain' 'Barbados' 'Belarus' 'Belgium'
 'Belize' 'Bosnia and Herzegovina' 'Brazil' 'Bulgaria' 'Cabo Verde'
 'Canada' 'Chile' 'Colombia' 'Costa Rica' 'Croatia' 'Cuba' 'Cyprus'
 'Czech Republic' 'Denmark' 'Dominica' 'Ecuador' 'El Salvador' 'Estonia'
 'Fiji' 'Finland' 'France' 'Georgia' 'Germany' 'Greece' 'Grenada'
 'Guatemala' 'Guyana' 'Hungary' 'Iceland' 'Ireland' 'Israel' 'Italy'
 'Jamaica' 'Japan' 'Kazakhstan' 'Kiribati' 'Kuwait' 'Kyrgyzstan' 'Latvia'
 'Lithuania' 'Luxembourg' 'Macau' 'Maldives' 'Malta' 'Mauritius' 'Mexico'
 'Mongolia' 'Montenegro' 'Netherlands' 'New Zealand' 'Nicaragua' 'Norway'
 'Oman' 'Panama' 'Paraguay' 'Philippines' 'Poland' 'Portugal'
 'Puerto Rico' 'Qatar' 'Republic of Korea' 'Romania' 'Russian Federation'
 'Saint Kitts and Nevis' 'Saint Lucia' 'Saint Vincent and Grenadines'
 'San Marino' 'Serbia' 'Seychelles' 'Singapore' 'Slovakia' 'Slovenia'
 'South 

In order to convert the numpy array to a Python list, we will use the numpy `tolist()` method. In case the country list is not alphabetically sorted, we can use the `sorted()` function.


In [23]:
country_list = country_array.tolist() # Convert a numpy array to a list
sorted_country_list = sorted(country_list)
country_list  # The list starts and finishes with []
#  we can print the list within the print function:      print(f'The list of countries is:{data_countries_list}')

['Albania',
 'Antigua and Barbuda',
 'Argentina',
 'Armenia',
 'Aruba',
 'Australia',
 'Austria',
 'Azerbaijan',
 'Bahamas',
 'Bahrain',
 'Barbados',
 'Belarus',
 'Belgium',
 'Belize',
 'Bosnia and Herzegovina',
 'Brazil',
 'Bulgaria',
 'Cabo Verde',
 'Canada',
 'Chile',
 'Colombia',
 'Costa Rica',
 'Croatia',
 'Cuba',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'Dominica',
 'Ecuador',
 'El Salvador',
 'Estonia',
 'Fiji',
 'Finland',
 'France',
 'Georgia',
 'Germany',
 'Greece',
 'Grenada',
 'Guatemala',
 'Guyana',
 'Hungary',
 'Iceland',
 'Ireland',
 'Israel',
 'Italy',
 'Jamaica',
 'Japan',
 'Kazakhstan',
 'Kiribati',
 'Kuwait',
 'Kyrgyzstan',
 'Latvia',
 'Lithuania',
 'Luxembourg',
 'Macau',
 'Maldives',
 'Malta',
 'Mauritius',
 'Mexico',
 'Mongolia',
 'Montenegro',
 'Netherlands',
 'New Zealand',
 'Nicaragua',
 'Norway',
 'Oman',
 'Panama',
 'Paraguay',
 'Philippines',
 'Poland',
 'Portugal',
 'Puerto Rico',
 'Qatar',
 'Republic of Korea',
 'Romania',
 'Russian Federation',
 'Saint 

In [25]:
#Find the length of the list
len(country_list)

101

We have looked at how to explore categorical data. Now let's look at numeric data. We can get the mean of a column using the `mean()` function

In [28]:
case_numbers = df_raw['cases_no']
case_numbers

0         21
1         16
2         14
3          1
4          9
        ... 
27835      6
27836    152
27837     21
27838    NaN
27839      7
Name: cases_no, Length: 27840, dtype: object

In [30]:
cases_mean = df_raw['cases_no'].mean()

TypeError: can only concatenate str (not "int") to str

You should have received an error here, and the key reason for this can be seen at the bottom fo the printout of the variable `cases_no`. This column is of the data type `object`. This typically denotes a mixture of different data, both numeric and non-numeric. As a result, pandas cannot work out a mean. We will focus on handling missing data in the next section.
 


## 3. Handling missing data
In the above printout of `cases_no`, we can see the code "NaN" in some of the rows, which is a special numeric value in pandas which means "Not A Number". When Pandas reads in the file, it identifies empty values, as well as certain strings (for example `#N/A` or `None`), as missing data and replaces this with the NaN code. However, especially when data has not yet been cleaned, there may be many different ways that a missing value is indicated, such as the string `Null` or a specific numeric code. There are many ways of handling missing data, but the first step in all of them is to identify the missing data and make sure they are labelled as such by pandas. In this section, we will identify all of the missing values and prepare the format of our DataFrame to be ready for analysis.

Steps for working with missing data:
* Identify missing data
* Deal with missing data
* Correct data format

### 3.1 Identifying missing data
Pandas represents missing data with a value of `NaN`. We can check whether a value is missing or not using the `isnull()` or `isna()` function. The opposite function to detect non-missing values is `notnull()` or `notna()`. The output from these functions are boolean (i.e. `True` or `False`.  

The following code will check every cell in our DataFrame and tell us how many have missing values. 

In [31]:
df_raw.isnull().value_counts() # Indicates with True that there are Nulls and counts the values that are numbers 

country  year   sex    age    cases_no  population  HDI for year   gdp_for_year ($) 
False    False  False  False  False     False       True          False                 16357
                                                    False         False                  7218
                              True      False       True          False                  3115
                                                    False         False                  1150
dtype: int64

The output shows all of the combinations with missing data. Two columns - `cases_no` and `HDI for year` - have some mimssing values because the `isnull()` test has returned `True`. 

In [32]:
missing_data = df_raw.isnull() # find data that are null and return 'True' if is null and 'False' if it is not
missing_data.head(20)

country   year    sex    age  cases_no  population  HDI for year  \
0     False  False  False  False     False       False          True   
1     False  False  False  False     False       False          True   
2     False  False  False  False     False       False          True   
3     False  False  False  False     False       False          True   
4     False  False  False  False     False       False          True   
5     False  False  False  False     False       False          True   
6     False  False  False  False     False       False          True   
7     False  False  False  False     False       False          True   
8     False  False  False  False     False       False          True   
9     False  False  False  False      True       False          True   
10    False  False  False  False      True       False          True   
11    False  False  False  False      True       False          True   
12    False  False  False  False     False       False          True   
13    False  False  False  False     False       False          True   
14    False  False  False  False     False       False          True   
15    False  False  False  False     False       False          True   
16    False  False  False  False     False       False          True   
17    False  False  False  False     False       False          True   
18    False  False  False  False     False       False          True   
19    False  False  False  False     False       False          True   

     gdp_for_year ($)   
0                False  
1                False  
2                False  
3                False  
4                False  
5                False  
6                False  
7                False  
8                False  
9                False  
10               False  
11               False  
12               False  
13               False  
14               False  
15               False  
16               False  
17               False  
18               False  
19               False

What if we wanted to determine the number of missing values in each column? We can use a loop to get the number of missing values found in each column.

In [33]:
# This loop will go through each column in the data frame
for column in missing_data: # Select each column-name in the header and prints it 
    print(column)
    print(missing_data[column].value_counts())
    print('----')

country
False    27840
Name: country, dtype: int64
----
year
False    27840
Name: year, dtype: int64
----
sex
False    27840
Name: sex, dtype: int64
----
age
False    27840
Name: age, dtype: int64
----
cases_no
False    23575
True      4265
Name: cases_no, dtype: int64
----
population
False    27840
Name: population, dtype: int64
----
HDI for year
True     19472
False     8368
Name: HDI for year, dtype: int64
----
 gdp_for_year ($) 
False    27840
Name:  gdp_for_year ($) , dtype: int64
----


The length of the dataframe is 27840 and the column `cases_no` has 23575 rows where `isnull()` is False, indicating that there is numeric data, and it has 4265 values of True (27840 - 23575). 

The HDI column has more observations with missing values - 19472 in total. As we will not use this column for analysis, we can simply remove it.

#### Identifying non-numeric data in a column
Most of the missing values have already been identified and appropriately labelled by pandas, but since this column is not a numeric datatype, there must still be some non-numeric data stored in it. All the values are currently strings, so we can find all rows where the strings represent data that is non-numeric. We will use the method `str.isdigit()`, which will generate a set of Boolean values much like our test above with `isnull()` did. 

In [43]:
non_numeric = ( df_raw['cases_no'].str.isdigit() == False)
non_numeric.value_counts()

False    27820
True        20
Name: cases_no, dtype: int64

 Notice how this method has only found twenty values, far less than we found with the `isnull()` method. That is because `NaN` is considered a numeric value. However, these are the values that are causing all of the problems, and we need to find out what they are and handle them appropriately. 
 
We are using a trick in pandas called _boolean indexing_. The following code will only select rows in the column `cases_no` where the variable `non_numeric` is equal to `True`. Thus it will return only twenty rows, the ones that have non-numeric data. Then, we will use the method `unique()` again to remove duplicate values. 

In [45]:
df_raw['cases_no'][non_numeric].unique()

array(['Null', 'Unknown'], dtype=object)

So these twenty rows have two unique values: `Null` and `Unknown`. These two strings are not automatically identified and labelled as missing by pandas when reading in the file and now we can set these values as missing. We can use the `replace()` function to remove these values.

In [5]:
df_raw['cases_no'] = df_raw['cases_no'].replace('Null',np.nan)
df_raw['cases_no'] = df_raw['cases_no'].replace('Unknown',np.nan)
non_numeric = ( df_raw['cases_no'].str.isdigit() == False)
non_numeric.value_counts()

False    27840
Name: cases_no, dtype: int64

### Exercise 4
The output should indicate that `cases_no` no longer has non-numeric data. How can you tell that? Put your answer in the blank markdown cell below.

#### Answer

### Exercise 5
What datatype is `cases_no` now? Will the mean work?

In [ ]:
# Put your answer to Exercise 5 here

So we still need to have pandas convert this to a numeric type that will allow us to do more analysis with it.  

In [6]:
df_raw['cases_no'] = df_raw['cases_no'].apply(pd.to_numeric)
print(df_raw['cases_no'].dtype)
df_raw['cases_no']

float64


0         21.0
1         16.0
2         14.0
3          1.0
4          9.0
         ...  
27835      6.0
27836    152.0
27837     21.0
27838      NaN
27839      7.0
Name: cases_no, Length: 27840, dtype: float64

Does the mean now work?

In [7]:
cases_mean = df_raw['cases_no'].mean()
print(cases_mean)

217.80072171513478


### 3.1 Deal with missing data
Now that we have identified what data is missing, how should we handle that for our analysis? There are many options, and the right one will be based on the data and what assumptions that can be made on why the data is missing. For some of the more complex options, you should consult with a statistician before implementing.

* Drop data
  * Drop the whole row
  * Drop the whole column
* Replace data - often known as imputation
  * Replace it with the mean
  * Replace it with multiple guesses from the expected distribution (multiple imputation)
  * Replace it by the mode
  * Replace it based on other functions




####  Dropping a column with lots of missing data.
To remove a column we use the `pandas.DataFrame.drop()` method. The input argument should be the list of column names that you want to drop, and axis argument should be set to one to tell pandas that you are wanting to drop columns, not rows. 

We don't need the column `HDI for year` for our analysis, and since there are a lot of missing values there, we can just drop it from the data frame. 

To avoid writing over the original data_frame, we will save the output to a new variable.


In [8]:
df_no_hdi = df_raw.drop(['HDI for year'], axis = 1)                         
df_no_hdi

country  year     sex          age  cases_no  population  \
0          Albania  1987    male  15-24 years      21.0      312900   
1          Albania  1987    male  35-54 years      16.0      308000   
2          Albania  1987  female  15-24 years      14.0      289700   
3          Albania  1987    male    75+ years       1.0       21800   
4          Albania  1987    male  25-34 years       9.0      274300   
...            ...   ...     ...          ...       ...         ...   
27835      Belgium  2011  female  25-34 years       6.0      707535   
27836     Thailand  2016    male    75+ years     152.0     1124052   
27837  Netherlands  1998  female  15-24 years      21.0      934500   
27838      Grenada  2002  female   5-14 years       NaN       11760   
27839       Mexico  1988  female    75+ years       7.0      614000   

       gdp_for_year ($)   
0          2,156,624,900  
1          2,156,624,900  
2          2,156,624,900  
3          2,156,624,900  
4          2,156,624,900  
...                  ...  
27835    527,008,453,887  
27836    411,755,164,833  
27837    432,476,116,419  
27838        540,336,926  
27839    183,144,164,357  

[27840 rows x 7 columns]

Now we turn to the other variable in the data frame that had missing values `cases_no`. Now that all of the missing values are coded properly, we can identify all of the missing data with the `isnull()` function. 

In [9]:
missing_values = df_no_hdi['cases_no'].isnull()
missing_values.value_counts()

False    23555
True      4285
Name: cases_no, dtype: int64

To do a _complete case analysis_, we can simply use the `dropna()` function. Even though `cases_no` is the only column with missing data, we tell dropna to only consider it when decided which rows to drop. We use `axis=0` to tell it to drop rows not columns. 

In [10]:
df_no_missing_cases = df_no_hdi.dropna(axis=0, subset=['cases_no'])
df_no_missing_cases.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23555 entries, 0 to 27839
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   country             23555 non-null  object 
 1   year                23555 non-null  int64  
 2   sex                 23555 non-null  object 
 3   age                 23555 non-null  object 
 4   cases_no            23555 non-null  float64
 5   population          23555 non-null  int64  
 6    gdp_for_year ($)   23555 non-null  object 
dtypes: float64(1), int64(2), object(4)
memory usage: 1.4+ MB


Note that the new data frame only contains 23555 entries, which is how many non-missing values there are. Let's take a look at the actual data. 

In [64]:
df_no_missing_cases

country  year     sex          age  cases_no  population  \
0          Albania  1987    male  15-24 years      21.0      312900   
1          Albania  1987    male  35-54 years      16.0      308000   
2          Albania  1987  female  15-24 years      14.0      289700   
3          Albania  1987    male    75+ years       1.0       21800   
4          Albania  1987    male  25-34 years       9.0      274300   
...            ...   ...     ...          ...       ...         ...   
27834      Ukraine  2005  female  25-34 years     182.0     3380536   
27835      Belgium  2011  female  25-34 years       6.0      707535   
27836     Thailand  2016    male    75+ years     152.0     1124052   
27837  Netherlands  1998  female  15-24 years      21.0      934500   
27839       Mexico  1988  female    75+ years       7.0      614000   

       gdp_for_year ($)   
0          2,156,624,900  
1          2,156,624,900  
2          2,156,624,900  
3          2,156,624,900  
4          2,156,624,900  
...                  ...  
27834     86,142,018,069  
27835    527,008,453,887  
27836    411,755,164,833  
27837    432,476,116,419  
27839    183,144,164,357  

[23555 rows x 7 columns]

It's a bit odd that we have only 23,555 rows now but some of the rows in the spreadsheet still have their old index (27839). We can fix that by using `reset_index()`

In [11]:
df_no_missing_cases = df_no_missing_cases.reset_index()
df_no_missing_cases

index      country  year     sex          age  cases_no  population  \
0          0      Albania  1987    male  15-24 years      21.0      312900   
1          1      Albania  1987    male  35-54 years      16.0      308000   
2          2      Albania  1987  female  15-24 years      14.0      289700   
3          3      Albania  1987    male    75+ years       1.0       21800   
4          4      Albania  1987    male  25-34 years       9.0      274300   
...      ...          ...   ...     ...          ...       ...         ...   
23550  27834      Ukraine  2005  female  25-34 years     182.0     3380536   
23551  27835      Belgium  2011  female  25-34 years       6.0      707535   
23552  27836     Thailand  2016    male    75+ years     152.0     1124052   
23553  27837  Netherlands  1998  female  15-24 years      21.0      934500   
23554  27839       Mexico  1988  female    75+ years       7.0      614000   

       gdp_for_year ($)   
0          2,156,624,900  
1          2,156,624,900  
2          2,156,624,900  
3          2,156,624,900  
4          2,156,624,900  
...                  ...  
23550     86,142,018,069  
23551    527,008,453,887  
23552    411,755,164,833  
23553    432,476,116,419  
23554    183,144,164,357  

[23555 rows x 8 columns]

####  Replacing missing values
Replacing or inputting missing data can be a complex subject. Here we are only going to show the simplest option of replacing these missing values with a single value. One option is to replace the missing value with a mean.
In order to do that we make a copy of the DataFrame and then modify the specific cells identified as missing using the `loc[]` method which only modifies the rows that are missing in the `cases_no` column.

In [12]:
df_cases_replaced = df_no_hdi.copy()
df_cases_replaced.loc[missing_values,'cases_no'] = cases_mean
df_cases_replaced

country  year     sex          age    cases_no  population  \
0          Albania  1987    male  15-24 years   21.000000      312900   
1          Albania  1987    male  35-54 years   16.000000      308000   
2          Albania  1987  female  15-24 years   14.000000      289700   
3          Albania  1987    male    75+ years    1.000000       21800   
4          Albania  1987    male  25-34 years    9.000000      274300   
...            ...   ...     ...          ...         ...         ...   
27835      Belgium  2011  female  25-34 years    6.000000      707535   
27836     Thailand  2016    male    75+ years  152.000000     1124052   
27837  Netherlands  1998  female  15-24 years   21.000000      934500   
27838      Grenada  2002  female   5-14 years  217.800722       11760   
27839       Mexico  1988  female    75+ years    7.000000      614000   

       gdp_for_year ($)   
0          2,156,624,900  
1          2,156,624,900  
2          2,156,624,900  
3          2,156,624,900  
4          2,156,624,900  
...                  ...  
27835    527,008,453,887  
27836    411,755,164,833  
27837    432,476,116,419  
27838        540,336,926  
27839    183,144,164,357  

[27840 rows x 7 columns]

You can see that the next to last row has the mean instead of a specific number. 

Replace "NaN" with the integer average value in the "suicide_no" column.
As written above we will not replace by a mean in this example.
I just show the code to use it when appropriate.



Another option is to replace with the most frequent value, also known as the mode. This can be done with the `mode()` function. As there can be more than one mode, it will return a pandas Series. However, we can see here that there is only one mode, so we will replace using that value. 

In [13]:
cases_mode = df_no_hdi['cases_no'].mode()
print(f"The mode is {cases_mode}")
df_cases_replaced.loc[missing_values,'cases_no'] = cases_mode.iloc[0]
df_cases_replaced

The mode is 0    1.0
dtype: float64


country  year     sex          age  cases_no  population  \
0          Albania  1987    male  15-24 years      21.0      312900   
1          Albania  1987    male  35-54 years      16.0      308000   
2          Albania  1987  female  15-24 years      14.0      289700   
3          Albania  1987    male    75+ years       1.0       21800   
4          Albania  1987    male  25-34 years       9.0      274300   
...            ...   ...     ...          ...       ...         ...   
27835      Belgium  2011  female  25-34 years       6.0      707535   
27836     Thailand  2016    male    75+ years     152.0     1124052   
27837  Netherlands  1998  female  15-24 years      21.0      934500   
27838      Grenada  2002  female   5-14 years       1.0       11760   
27839       Mexico  1988  female    75+ years       7.0      614000   

       gdp_for_year ($)   
0          2,156,624,900  
1          2,156,624,900  
2          2,156,624,900  
3          2,156,624,900  
4          2,156,624,900  
...                  ...  
27835    527,008,453,887  
27836    411,755,164,833  
27837    432,476,116,419  
27838        540,336,926  
27839    183,144,164,357  

[27840 rows x 7 columns]

Now instead of the mean (217.8), the missing value is the penultimate row is replaced with teh mode (1.0)

## 4. Checking the data representation

We have previously corrected the issues with the `cases_no` column. Now let's take a look what issues are still remaining by using the `info()` function.

In [14]:
df_cases_replaced.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27840 entries, 0 to 27839
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   country             27840 non-null  object 
 1   year                27840 non-null  int64  
 2   sex                 27840 non-null  object 
 3   age                 27840 non-null  object 
 4   cases_no            27840 non-null  float64
 5   population          27840 non-null  int64  
 6    gdp_for_year ($)   27840 non-null  object 
dtypes: float64(1), int64(2), object(4)
memory usage: 1.5+ MB


We would expect age and gdp to be numeric, but they are not. Let's check the reason why by looking at whether they are numeric or not.


In [17]:
print(df_cases_replaced['age'].unique())

['15-24 years' '35-54 years' '75+ years' '25-34 years' '55-74 years'
 '5-14 years']


So age does not represent a numeric value in years, but rather it ia a _categorical_ variable representing age bands. We will convert this column to a categorical value. But if you see above, the values are not quite as sorted as we would like the category to be (look at where the 5-14 year old group is). So we need to tell pandas what order we want these age bands in.

In [26]:
age_bands_sorted = [
    '5-14 years','15-24 years','25-34 years', 
    '35-54 years','55-74 years','75+ years'  
    ]
df_cases_replaced['age'] = pd.Categorical(df_cases_replaced['age'], 
                                          categories=age_bands_sorted)
df_cases_replaced['age']

0        15-24 years
1        35-54 years
2        15-24 years
3          75+ years
4        25-34 years
            ...     
27835    25-34 years
27836      75+ years
27837    15-24 years
27838     5-14 years
27839      75+ years
Name: age, Length: 27840, dtype: category
Categories (6, object): ['5-14 years', '15-24 years', '25-34 years', '35-54 years', '55-74 years', '75+ years']

See how this variable now has a dtype of `category` and the categories are sorted so that the levels go up with age?

Now what is going on with the column representing gdp? First we look at how many values do not have numeric data.

In [20]:
gdp_is_numeric = (df_cases_replaced[' gdp_for_year ($) '].str.isdigit()==False)
gdp_is_numeric.value_counts()


True    27840
Name:  gdp_for_year ($) , dtype: int64

All of these are not numeric. Let's look at head again to see what is up here?

In [22]:
df_cases_replaced[' gdp_for_year ($) '].head()

0    2,156,624,900
1    2,156,624,900
2    2,156,624,900
3    2,156,624,900
4    2,156,624,900
Name:  gdp_for_year ($) , dtype: object

It is likely that the commas are causing trouble here. If we had spotted this before we read in the file, we could add the argument `thousands=','` to our `read_csv` call. 

Before we fix this though the leading and trailing spaces in the column are annoying, so let's get rid of those first. 

In [23]:
df_cases_replaced = df_cases_replaced.rename(columns={' gdp_for_year ($) ':'gdp_for_year_usd'})
df_cases_replaced.columns

Index(['country', 'year', 'sex', 'age', 'cases_no', 'population',
       'gdp_for_year_usd'],
      dtype='object')

Fixing the columns requires we do a string operation to the values.

In [24]:
# First replace the string comma ',' empty space''. Then typecasting the column to integers
df_cases_replaced['gdp_for_year_usd'] = df_cases_replaced['gdp_for_year_usd'].str.replace(',', '').astype(int)
df_cases_replaced['gdp_for_year_usd']

0          2156624900
1          2156624900
2          2156624900
3          2156624900
4          2156624900
             ...     
27835    527008453887
27836    411755164833
27837    432476116419
27838       540336926
27839    183144164357
Name: gdp_for_year_usd, Length: 27840, dtype: int64

Notice the dtype at the bottom of the printout which should now say `int64` instead of `object`.

### **Wonderful!**
#### Save the data

Now we have finally obtained the cleaned dataset with no missing values with all data in its proper format.
We can save this data set as it is now the 'clean' data set which is ready for processing. 


In [ ]:
df_cases_replaced.to_csv(r'who_suicide_statistics_clean_data.csv', index=False)

# **Data Operations**
Data operations can be performed through various built-in methods for faster data processing and analysis. A few methods are:
- Operations using Grouping
- Operations with Sorting
- Operations with Statistics 
- Operations with Functions (Standarisation and Normalisation)




##  Grouping
 Use  **.groupby('..')** Its a function of grouping the data. It involes a combination of **spliting** the object applying a function and **combine** the results **.groupby('column name, value')**

To select the data of a single country from the 101, I'll  group the data frame by 'country' and from that I will select the country of preference by using **.get_group('selection').
Lets select the data for the United Kingdom.

In [ ]:
country_grouped= data_frame.groupby(['country'])
UK_data=country_grouped.get_group('United Kingdom')
UK_data

We make another data frame by selecting the columns that we want to analyse 

In [287]:
UK_data_year_suicides= UK_data[['country','year', 'suicides_no']] # UK_data = country_grouped.get_group('United Kingdom'). If you want you could add another column as well
UK_data_year_suicides

country  year  suicides_no
26476  United Kingdom  1985          264
26477  United Kingdom  1985          915
26478  United Kingdom  1985          128
26479  United Kingdom  1985           62
26480  United Kingdom  1985          678
...               ...   ...          ...
26843  United Kingdom  2015          181
26844  United Kingdom  2015           18
26845  United Kingdom  2015           14
26846  United Kingdom  2015            6
26847  United Kingdom  2015            4

[369 rows x 3 columns]

To find the total suicides of each year, group two columns together and use the  method **.sum()**


In [ ]:
UK_suicides_per_year =UK_data_year_suicides.groupby(['country','year'],as_index=False).sum()
UK_suicides_per_year.head(10)

## Sorting values 
To sort values either according to their size or alphabetical order we use  the method **.sort_values**

In [ ]:
UK_suicides_per_year_sorted=UK_suicides_per_year.sort_values('suicides_no') # The default is ascending order
UK_suicides_per_year_sorted

## Operations with Statistics 
The methods **.max()**, **.min()**,**.mean()**,**.std()** used to find the maximum, minimum, mean and standard deviation at data. The **.describe** function shows all the statistics.

In [ ]:
UK_suicides_per_year['suicides_no'].max() # Select from the data frame, the column that we are interested and apply the .max() method

In [ ]:
# Use the print() function to print the numbers that you calculate with the .max(), .min(). mean() .std() and all the statistical methods
print('The maximum number of cases in the UK is', UK_suicides_per_year['suicides_no'].max())
print('The minimum number of cases in the UK is',UK_data_year_suicides['suicides_no'].min())
print('The standard deviation of cases in the UK is', UK_suicides_per_year['suicides_no'].std())

## Investigate another country's data and create another sub-data 

Let's repeat what we have calculated and presented for the UK for another country. You could choose any country from the list of the initial data frame. I'm choosing Greece here and I repeat the code above:

In [ ]:
## Investigate another data set 
country_grouped= data_frame.groupby(['country'])
Greece_data=country_grouped.get_group('Greece')
Greece_data

In [ ]:
Greece_data_year_suicides= Greece_data[['country','year', 'suicides_no']] # country_grouped.get_group('Greece').
Greece_data_year_suicides

In [ ]:
Greece_suicides_per_year =Greece_data_year_suicides.groupby(['country','year'],as_index=False).sum()
Greece_suicides_per_year.head(10)

In [ ]:
Greece_suicides_per_year_sorted=Greece_suicides_per_year.sort_values('suicides_no') # The default is ascending order
Greece_suicides_per_year_sorted

In [ ]:
print('The maximum number of cases in Greece is', Greece_suicides_per_year['suicides_no'].max())
print('The minimum number of cases in Greece is',Greece_data_year_suicides['suicides_no'].min())
print('The standard deviation of cases in Greece',Greece_suicides_per_year['suicides_no'].std())



## Combine / Merge the data sets 

Another useful operation is to combine together data sets. Pandas is a powerful library and gives a multifaced approach to combining separate datasets. 
With pandas, you can **merge** , **join** and **concatenate** your datasets, allowing you to unify them , understand and interpret them. 

![merge](./fig/fig_merge.jpg)


If you use the **merge() function** , the default is an inner join, which in most of the cases result in a DataFrames which are shorter. With merge(), you also have control over which column(s) to join on. Let’s say that you want to merge both entire datasets, but only on 'year'  since the combination of the two will yield a unique value for each row. To do so, you can use the on parameter:

In [ ]:
UK_suicides_per_year.shape

In [ ]:
Greece_suicides_per_year.shape

In [ ]:
UK_Greece_merged=pd.merge(UK_suicides_per_year, Greece_suicides_per_year,on='year') # Merges the two datasets (one next to the other) and keep common column the 'year' column
UK_Greece_merged

Depending on the datasets you can join on 'left' or on the 'right but in this example it doesn't mean anything: The indication is with the **how** parameter like below: 

In [ ]:
df1_merge=pd.merge(UK_suicides_per_year, Greece_suicides_per_year,how='right', on='year')
df1_merge 

## Use the .join() method : Combine Data on a Column or index
The merge () is a module function (which means that you write the DataFrames as arguments). The .join() is method.  This enables you to specify only one DataFrame, which will join the DataFrame you call .join() on. 

Here I'll create on data frame (Two_countries_combined) which I'll assign it  the Greek results and then I'll join to it the UK DataFrame.

In [ ]:
one_country= Greece_suicides_per_year

In [ ]:


two_countries_combined=one_country.join(UK_suicides_per_year.set_index(["year"]), on=["year"], how= "inner", lsuffix="_x", rsuffix="_y") # It is important to set the index at th common column. The lsuffix sets the first data Frame on the left
two_countries_combined

In [ ]:
two_countries_combined2=one_country.join(UK_suicides_per_year.set_index(["year"]), on=["year"], lsuffix="_left") # In this datasets if you dont use how "inner" it doesnt matter 
two_countries_combined2

## Use the concat() method : Combine data Across Rows or Columns 
Concatenation is a bit different from the merging techniques with merge() function and .join() method where you would specify the merging (inner or outer), and depending on the type of merge you might lose rows and information. 

With the concatenation the  datasets are stitched together along wither the row axis or the column axis.

If you us the following code that concatenates the dataframes (df1 and df2)  with no parameters the default results will look like the picture:

**df_concatenated = pandas.concat([df1, df2])**



![concatenate1](./fig/fig1_concat.jpg)

This example assumes that your column names are the same.  If your column names are different while concatenating along rows (axis 0), then by default the columns will also be added, and NaN values will be filled in as applicable.

![concatenate2](./fig/fig2_concat.jpg)

If you want to accomplish concatenation along columns instead, you’ll use a concat() call like you did above, but you’ll also need to pass the axis parameter with a value of 1 or "columns":

**df_concatenated = pandas.concat([df1, df2], axis='columns")**  or **df_concatenated = pandas.concat([df1, df2], axis=1)**



![concatenate3](./fig/fig3_concat.jpg)

Lets concatenate our datasets and view them 

In [ ]:
df1_concatenated=pd.concat([UK_suicides_per_year, Greece_suicides_per_year])
df1_concatenated

This results to the datasets being one underneath the other. They are concatenated by default on the "row axis" or axis=0. If you look carefully the indexes are different for each data sets (start from 0-30 for the first dataframe and from 0-30 for the second one)

In [ ]:
df2_concatenated=pd.concat([UK_suicides_per_year, Greece_suicides_per_year], axis =1)
df2_concatenated

You see the data frame 1 next to the other.

By default, a concatenation results in a **set union**, where all data is preserved. You’ve seen this with merge() and .join() as an outer join, and you can specify this with the join parameter.



## **Visualise the data**

Where you are investicating your data of prepare to publish your findings, visualisation is an essential tool. There are various libraries that you can use in Python to visualise your data most popular are Matplotlib, and Seaborn. Here we use the Matplotlib library 

In [ ]:
%matplotlib inline  

  **%matplotlib** is a magic function, if you use it just it displays the figure as a pup up. Using with inline, the figure is inside the noteboook 

In [ ]:
df1_merged=pd.merge(UK_suicides_per_year, Greece_suicides_per_year,how='right', on='year') #
df1_merged 



We can change the column names to be more specific or use the .join() method instead which have arguments lsuffix and rsuffix to specify it. 
Lets change the header as we have seen before at paragraph 1. 


In [ ]:
headers=['UK','year','suicides_no_UK','Greece','suicides_no_Greece']
df1_merged.columns=headers
df1_merged.head()

Now we are ready to plot. Firts import the pyplot module of the matplotlib library as plt.
then use the .plot() function at our dataframe. The arguments start with x column and y column/s

In [ ]:

import matplotlib.pyplot as plt
df1_merged.plot(x='year', y=['suicides_no_UK','suicides_no_Greece'])



The default plot is line plot. There are different other kinds of plots and we need to specify them in the arguments these are : bar, area, barh=horizontal bars, hist= histogramms, box, pie= piecharts, line, scatter. 

In [ ]:
df1_merged.plot(kind= 'bar', x='year', y=['suicides_no_UK','suicides_no_Greece'])

## Operations with Functions Statistical Functions
### Example Statistical Functions of **Normalization** and **Standarization** 
Data is usually collected from different sources in different formats.
Normalization and standardization are the processes of transforming data into a common format, allowing the researcher to make the meaningful comparison.

**Normalization** is the process of transforming values of several variables into a similar range. Typical normalizations include scaling the variable from 0 to 1, scaling the variable so the variance is 



### **Example**

To demonstrate normalization, let's say we want to scale the columns , "suicides_no_UK" and "suicides_no_Greece" at df1_merged data frame.
We would like to normalize those variables so their value ranges from 0 to 1
We replace original value by (original value)/(maximum value)

**Step 1)** Create a function of normalization 

In [ ]:
def normalize(number):
    return (number)/number.max()

**Step 2)** Apply the function to the selected column 

In [ ]:
suicides_UK_normalised=normalize(df1_merged['suicides_no_UK'])
suicides_UK_normalised

In [ ]:
suicides_Greece_normalised=normalize(df1_merged['suicides_no_Greece'])
suicides_Greece_normalised

### **Data Standardization**
Data standardization is also a term for a particular type of data normalization where we subtract the mean and divide by the standard deviation. The data is transformed into a common format with  the mean to be 0 and the variable changes from negative  to positive around the mean. 

In [ ]:
def standarize_test(number):
    return (number-number.mean())/number.std()

In [ ]:
standarize_test(UK_suicides_per_year['suicides_no'])

## Visualise and compare
 I want to make a comparison between the normalised values fo both countries and I use again the merged data.

We can add the two columns of normalised to the df1_merged dataset and create another datasets which includes the statistics. Or we can create a new data set.

Lets go with the first option:

In [ ]:
df1_merged['UK_suicides_normalised']= suicides_UK_normalised # Add a new column at the merged dataframe and add the series of normalisation data as we calculated above
df1_merged['Greece_suicides_normalised']= suicides_Greece_normalised
df1_merged.head()

We can now plot the normalisation data and get insightful information comparing the two countries.

In [ ]:
df1_merged.plot(kind= 'bar', x='year', y=['UK_suicides_normalised','Greece_suicides_normalised'])# we can add in the dataframe the plot()function

You see that in general the % of the cases in the UK was higher until 2011. After 2011 the % cases in Greece were higher and it is due to the economic crisis starting fro 2010 and have the greatest impact to the society after 2011. 
With these kind of comparative plots you can visualise and give insightful information about political and socio-economic of different countries in the 101 one presented in this data. 


## Grouping plots
You can group similar plots in a single figure using subplots. The matplotlib.pyplot.figure() which is shortened to plt.figure() creates a space into which we will place all our plots. The parameter figsize tells Python how big to take this space. Each subplot is placed into the figure using its add_subplot method. The add_subplot method takes 3 parameters. The first denotes how many total rows of subplots there are, the second parameter refers to the total number of subplot columns, and the final parameter denotes which subplot your variable is referencing (left-to-right, top-to-bottom). Each subplot is stored in a different variable (axes1, axes2, axes3). Once a subplot is created, the axes can be titled using the ax1.set_title() or set_xlabel() command (or set_ylabel()). Here are our four plots side by side:

In [ ]:
fig=plt.figure(figsize=(12.0, 8.0))


ax1= fig.add_subplot(2,2,1)
#Here we plot bar using the plt.bar() function and it is set at the ax1 which is axes1 subplot variable
plt.title('UK suicides') 
plt.bar(UK_data['year'],
        UK_data['suicides_no'],
        color='blue')


ax2= fig.add_subplot(2,2,2)
ax2.set_title('GDP in billions USD')
ax2.plot(UK_data['year'],UK_data['gdp_for_year_usd']/10**9, color='purple' )


ax3= fig.add_subplot(2,2,3)

plt.title('Greece suicides')
plt.bar(Greece_data['year'],
        Greece_data['suicides_no'],
        color='orange')


ax4= fig.add_subplot(2,2,4)
ax4.set_title('GDP in billions USD')
ax4.plot(Greece_data['year'],Greece_data['gdp_for_year_usd']/10**9, color='green' )


I want to plot the GDP per capita over the years for both countries
I need to calculate it by dividing the two columns and add the result as an extra column.

Lets plot the sum of the suicides over the years for all countries (101)over the years.  

We could write it in one line: From the dataframe, group by the countries column, summarise and sort the values of suicides and plot the suicides_no

In [ ]:


plt.figure(figsize=(16,8))
data_frame.groupby('country').sum().sort_values(by='suicides_no',ascending=False)[['suicides_no']][:101].plot(kind='bar',figsize=(16,8),title='Sum of suicides by country during 1985-2016')
plt.ylabel("Sum of suicides")
plt.xlabel("Country")

In [ ]:
# Here it plots the top 20 countries in suicides 
plt.figure(figsize=(14,6))
data_frame.groupby('country').sum().sort_values(by='suicides_no',ascending=False)[['suicides_no']][:20].plot(kind='bar',figsize=(16,8),color='purple', title='Sum of suicides of the top 20 countries during 30 years 1985-2016')
plt.ylabel("Sum of suicides")




## Exercise 
Calculate the GDP/capita of the countries and visualise the cases according to GDP/capita.



# Convert to GDP/capita

### Thank you for completing this lab!

## Author
##  Mary Tziraki


                                                                                     

## <h3 align="center"> © Health+Bioscience IDEAS 2022. All rights reserved. <h3/>
